### PR TITLE
feat(community): add OpenSSF Scorecard source — security health checks for any GitHub repo

### PR DESCRIPTION
**Pull request by athul-2003**
## Summary

Adds two queryable SQL tables over the free, no-auth [OpenSSF Scorecard API](https://api.securityscorecards.dev):

- **`scorecard.checks`** — up to 18 security health checks per repo, each scored 0–10 with a plain-English reason
- **`scorecard.project`** — aggregate score, scored date, commit SHA, and Scorecard tool version (data freshness metadata)

Checks include: Code-Review, Token-Permissions, Branch-Protection, Pinned-Dependencies, SAST, Signed-Releases, Fuzzing, CI-Tests, Contributors, Dependency-Update-Tool, and more.

> **Coverage note:** Results are available for public GitHub repositories tracked by OpenSSF Scorecard. Repositories not yet indexed return a 404.

## Why this source?

- **Zero auth, zero config** — no API key needed; works for any public GitHub repo by owner + repo name
- **Novel DSL pattern** — uses `{{filter.owner}}` and `{{filter.repo}}` path template substitution to inject filter values directly into the URL path (`/projects/github.com/{owner}/{repo}`)
- **Two tables, two use cases** — `checks` for per-check detail, `project` for aggregate score and data freshness
- **Genuinely actionable** — tells maintainers exactly which security practices to fix, not just that something is wrong

## Example queries

```sql
-- All checks, worst first (most improvement needed at top)
SELECT check_name, score, reason
FROM scorecard.checks
WHERE owner = 'expressjs' AND repo = 'express'
ORDER BY CASE WHEN score = -1 THEN 999 ELSE score END ASC

-- Only the critical gaps (excluding N/A checks)
SELECT check_name, score, reason
FROM scorecard.checks
WHERE owner = 'django' AND repo = 'django'
  AND score >= 0 AND score < 5

-- Quick health summary
SELECT
  COUNT(*) AS total_checks,
  SUM(CASE WHEN score >= 8 THEN 1 ELSE 0 END) AS passing,
  SUM(CASE WHEN score >= 0 AND score < 5 THEN 1 ELSE 0 END) AS critical
FROM scorecard.checks
WHERE owner = 'expressjs' AND repo = 'express'

-- Check when the score was last computed and which commit was evaluated
SELECT date, aggregate_score, repo_commit, scorecard_version
FROM scorecard.project
WHERE owner = 'expressjs' AND repo = 'express' 
```

## Validation

- `coral source lint`: ✅ passed
- `coral source add`: ✅ connected, 2 tables (`checks`, `project`), 2/2 test queries passed
- `scorecard.checks` tested against: `expressjs/express` (18 checks), `django/django`, `rust-lang/rust`
- `scorecard.project` tested: returns 1 row with `date`, `aggregate_score`, `repo_commit`, `scorecard_version`, `scorecard_commit`
- 404 case (repo not in Scorecard DB): returns `Source resource was not found (404)`
- Column types: `check_name` Utf8, `score` Int64 (-1 or 0–10), `reason` Utf8, `documentation_url` Utf8, `aggregate_score` Float64, `date` Utf8

### CODE REVIEW

File: sources/community/scorecard/README.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+ORDER BY CASE WHEN score = -1 THEN 999 ELSE score END ASC
+```
+
+```sql
+-- Only the failing checks (score < 5)
+SELECT check_name, score, reason
+FROM scorecard.checks
+WHERE owner = 'django'
+  AND repo  = 'django'
+  AND score < 5
```
[the remark this answers is not in the document]
Comment (athul-2003) [reply sources/community/scorecard/README.md]: The issue is fixed.

[the remark this answers is not in the document]
Comment (athul-2003) [reply sources/community/scorecard/README.md]: Fixed — added `AND score >= 0` to exclude -1 (N/A) checks from the failing-checks example.

**Review by jsummerfield — CHANGES_REQUESTED** [review]
Findings:

- Medium: `sources/community/scorecard/manifest.yaml:5` and `sources/community/scorecard/README.md:3` overstate the source as working for any public GitHub repository. The Scorecard REST API serves published/precomputed results; projects that are not included can return 404 rather than a score. Please reword this as public GitHub repositories tracked by OpenSSF Scorecard, and consider `allow_404_empty: true` if an untracked repository should behave as an empty result instead of a query error. OpenSSF documents the REST API behavior here: https://github.com/ossf/scorecard#scorecard-rest-api

- Medium: `sources/community/scorecard/README.md:8` gives the wrong install path: `sources/scorecard/manifest.yaml`. The PR adds `sources/community/scorecard/manifest.yaml`, so the first-run command fails from the repo root. The validation block repeats the wrong path at `README.md:108`.

- Medium: `sources/community/scorecard/manifest.yaml:46` only exposes the `checks` array and drops the top-level result metadata: `date`, aggregate `score`, `repo.commit`, and `scorecard.version` / `scorecard.commit`. Because the README tells users the data is cached and updated weekly, users need a way to query when the score was produced and which commit was scored. Please add a small summary/project table or otherwise expose that metadata.

- Low: `sources/community/scorecard/manifest.yaml:29` lists known checks but omits checks returned by the live API for the README's own example repository, including `Dependency-Update-Tool`, `CI-Tests`, and `Contributors`. Please make the list explicitly non-exhaustive or bring it up to date.

Thanks for adding this. The endpoint shape and no-auth test are promising; the fixes above should make the source less surprising for first-time users.

**Comment by athul-2003** [comment]
@jsummerfield 

Thanks for the thorough review! All four points addressed in the latest commit:

- "any public GitHub repository" → reworded to "repositories tracked by OpenSSF Scorecard" in both the manifest description and README, with an explicit note that unindexed repos return a 404
- Wrong install path → fixed in both the Install section (sources/community/scorecard/manifest.yaml) and the Validation block
- Missing result metadata → added scorecard.project table exposing date, aggregate_score, repo_commit, scorecard_version, and scorecard_commit using rows_path: [] on the same endpoint. Verified locally: returns 1 row with correct values
- Non-exhaustive checks list → added CI-Tests, Contributors, Dependency-Update-Tool; marked the list as explicitly non-exhaustive with a note that the set varies by repository

**Review by jsummerfield — CHANGES_REQUESTED** [review]
Findings:

- Medium: `sources/community/scorecard/manifest.yaml:38`, `sources/community/scorecard/manifest.yaml:40`, `sources/community/scorecard/README.md:45`, and `sources/community/scorecard/README.md:83` still turn Scorecard numbers into severity language like “critical”, “poor”, “fair”, and “good”. OpenSSF describes Scorecard as a heuristic, opinionated signal with false positives/negatives, and its scoring docs define 0-10 / aggregate weighting without those severity bands. Please soften this to language like “low-scoring checks” or “areas Scorecard flags for review” rather than presenting `< 5` as “critical security gaps”. Docs: https://github.com/ossf/scorecard/blob/main/README.md#what-is-scorecard and https://github.com/ossf/scorecard/blob/main/README.md#aggregate-score

The earlier tracked-repository wording, install path, summary metadata table, and non-exhaustive check list are much improved. This last wording change should keep the source from overstating what the upstream API proves.

**Comment by athul-2003** [comment]
@jsummerfield 

Thanks for the review.

I’ve updated the wording across both `manifest.yaml` and `README.md` to better align with OpenSSF’s guidance that Scorecard is a heuristic signal and not a strict severity classification.

Changes include:

* Replaced terms like `critical`, `poor`, and `good`
* Switched to neutral wording such as `low-scoring checks` and `areas Scorecard flags for review`
* Updated score descriptions from severity-style labels to neutral ranges (`highest`, `high`, `mid-range`, `low`, `lowest`)
* Added disclaimer notes linking to the official OpenSSF scoring documentation

**Review by jsummerfield — CHANGES_REQUESTED** [review]
Findings

- Medium: `sources/community/scorecard/manifest.yaml:52` and `:138` do not set `allow_404_empty: true`, even though an unindexed repository is a normal OpenSSF Scorecard API outcome. I verified a missing repo returns HTTP 404 with an empty body, and Coral currently turns that into a query error. Please make both tables return empty rows for that documented case. Docs: https://github.com/ossf/scorecard-webapp/blob/main/openapi.yaml

- Medium: `sources/community/scorecard/manifest.yaml:80` omits the API's `details` field from `checks`. The summary `reason` is useful, but `details` is where Scorecard carries the actionable evidence such as workflow files and warnings. Please add a nullable `details` column, preferably `Json` from `[details]`, and consider exposing `documentation.short` too. Docs: https://github.com/ossf/scorecard-webapp/blob/main/openapi.yaml

- Low: `sources/community/scorecard/manifest.yaml:143` exposes `date` as `Utf8`, but live API values are ISO-8601 timestamps like `2026-05-25T21:42:09Z`. Since this is the main freshness field, please expose it as a `Timestamp` with `format_timestamp`.

- Low: `sources/community/scorecard/README.md:22` and `manifest.yaml:73` still describe score `0` as a “critical gap”. Scorecard scores are heuristic signals and checks have separate risk levels, so this overstates the meaning. Please use wording like “lowest score” or “check not satisfied.” Docs: https://github.com/ossf/scorecard#aggregate-score and https://github.com/ossf/scorecard/blob/main/docs/checks.md

Thanks for continuing to refine this. The core endpoint, path templating, project metadata, and live `expressjs/express` queries look good; the remaining changes are about making the source robust and useful in the normal “not indexed yet” and remediation workflows.

**Comment by athul-2003** [comment]
@jsummerfield 

Thanks for the review — all requested fixes have been addressed.

Changes made:

* Added `allow_404_empty` to both `checks` and `project` response blocks so unindexed repositories now return empty results instead of errors
* Added missing `details` (`Json`) and `documentation_short` (`Utf8`) columns to the `checks` table
* Updated `date` from `Utf8` to proper `Timestamp` parsing using `format_timestamp` with `input: iso8601`
* Removed remaining “critical gap” wording and replaced it with neutral `lowest/highest` terminology, along with the heuristic disclaimer

File: sources/community/scorecard/README.md [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+> **Note:** The Scorecard API serves precomputed results for repositories already indexed by OpenSSF. Repositories not yet tracked return a 404. See the [Scorecard REST API docs](https://github.com/ossf/scorecard#scorecard-rest-api) for details on coverage.
+
+## Tables
+
+| Table | Required filters | Purpose |
+|---|---|---|
+| `scorecard.checks` | `owner`, `repo` | Per-check scores (0–10), reasons, actionable details, and doc links |
+| `scorecard.project` | `owner`, `repo` | Aggregate score, scored timestamp, commit SHA, tool version |
+
+Both tables return empty rows (not an error) when the repository is not yet indexed by OpenSSF Scorecard.
```
[the remark this answers is not in the document]
Comment (athul-2003) [reply sources/community/scorecard/README.md:33]: updated this with the changes

**Comment by athulkrishnan-h** [comment]
@jsummerfield

Follow-up on the `date` column type change (Low finding from the latest round):

The `format_timestamp` with `input: iso8601` has now been applied — the `date` column in `scorecard.project` is correctly typed as `Timestamp` in the latest commit (`b8cea3a`).

One thing I ran into: the initial attempt used a flat `path:` key alongside `kind: format_timestamp`, which failed DSL schema validation. The correct syntax requires a **nested `expr` block** wrapping the path:

```yaml
expr:
  kind: format_timestamp
  input: iso8601
  expr:
    kind: path
    path:
      - date
```

I found this pattern by reading the existing `sources/community/k8s/manifest.yaml` in this repo, which uses the same construct for `creationTimestamp`. Once I matched that structure, both `coral source add` and both test queries passed locally.

The `scorecard.project` table now correctly returns `date` as a `Timestamp` with ISO 8601 input parsing. All other fixes from previous rounds remain in place (`allow_404_empty`, `details` + `documentation_short` columns, neutral score wording, correct install path, `score >= 0` filter guidance).

**Review by jsummerfield — APPROVED** [review]
Findings: no blocking findings.

I rechecked the current head and the previous blockers are resolved: unindexed repositories now return empty rows, the checks table exposes the actionable details/documentation fields, the project table exposes freshness metadata with a parsed timestamp, and the Scorecard wording is framed as a heuristic signal rather than a severity verdict.

I also verified `coral source lint`, `coral source add`, `coral source test`, representative live queries against `expressjs/express`, and the empty 404 path for an unindexed repository. Thanks for iterating on this; this is ready for maintainer merge.
